core/types/schema.go

```diff
@@ -804,3 +742,228 @@ func (m *Modifier) Clean() error {
 func (m Modifier) String() string {
 	return string(m)
 }
+
+type Procedure struct {
+	// Name is the name of the procedure.
+	// It should always be lower case.
+	Name string `json:"name"`
+
+	// Parameters are the parameters of the procedure.
+	Parameters []*ProcedureParameter `json:"parameters"`
+
+	// Public is true if the procedure is public.
+	Public bool `json:"public"`
+
+	// Modifiers are the modifiers of the procedure.
+	Modifiers []Modifier `json:"modifiers"`
+
+	// Body is the body of the procedure.
+	Body string `json:"body"`
+
+	// Returns is the return type of the procedure.
+	Returns *ProcedureReturn `json:"returnTypes,omitempty"`
+	// Annotations are the annotations of the procedure.
+	Annotations []string `json:"annotations,omitempty"`
+}
+
+func (p *Procedure) Clean() error {
+	params := make(map[string]struct{})
+	for _, param := range p.Parameters {
+		err := param.Clean()
+		if err != nil {
+			return err
+		}
+
+		_, ok := params[param.Name]
+		if ok {
+			return fmt.Errorf(`duplicate parameter name: "%s"`, param.Name)
+		}
+
+		params[param.Name] = struct{}{}
+	}
+
+	if p.Returns != nil {
+		err := p.Returns.Clean()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return cleanIdent(&p.Name)
+}
+
+// IsView returns true if the procedure has a view modifier.
+func (p *Procedure) IsView() bool {
+	for _, m := range p.Modifiers {
+		if m == ModifierView {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ProcedureReturn holds the return type of a procedure.
+// EITHER the Type field is set, OR the Table field is set.
+type ProcedureReturn struct {
+	Types []*DataType  `json:"returnTypes,omitempty"`
+	Table []*NamedType `json:"returnTable,omitempty"`
+}
+
+func (p *ProcedureReturn) Clean() error {
+	if p.Types != nil {
+		for _, t := range p.Types {
+			return t.Clean()
+		}
+	}
```

Comment:
It's required that only one (or neither) of Type or Table are set.  Maybe Clean should error if they both are?  Or leave that to other systems?  While RPC conversions helpers can catch it, is there logic else where in transaction execution that would?